Better build process. Fix #1238, #1232

tnajdek opened the pull request:
This is a new, re-written build process to make it less cumbersome. Here are the major changes:

* Remove gulp, replace with custom scripts
* Symlink entire dirs where possible (#1232)
* Significantly speed up subsequent builds (macOS and Linux only) (#1238)
* Watch process now observes new/removed files, not only changed. Should now remove any temporary/editor backup etc. files as soon as these disappear from the source 
* Better console output during builds

dstillman commented:
This is great. Just for house style, can we rename trailing abbreviations to be capitalized (`getJs` -> `getJS`) and add spaces after statements (`if (`, `while (`, `for (`) to differentiate from functions?

dstillman commented on scripts/utils.js:
```diff
@@ -0,0 +1,123 @@
+	return path.relative(path.join(ROOT, dirName), path.join(ROOT, f));
+}
+
+const formatDirsforMatcher = dirs => {
```
Capital `F`

dstillman commented:
And should probably log a warning for earlier Node versions, so people know they're getting an suboptimal experience.

dstillman commented on scripts/utils.js:
```diff
@@ -0,0 +1,123 @@
+	if('isError' in global && global.isError) {
+		return;
+	}
+	console.log(`${colors.blue(`[${operation}]`)} ${sourcefile} -> ${outfile}`);
```
I think we can remove the ` -> ${outfile}`. All it's ever adding is `build/` at the beginning, right? Easier to read without that.

dstillman commented:
The `npm i` on Travis is [failing](https://travis-ci.org/zotero/zotero/jobs/244507026#L179) with `Cannot find module 'array-union'`. I actually saw that locally, but then I think I cleared `node_modules` to fix it. But clearing the cache on Travis doesn't seem to be doing the trick.

tnajdek commented:
Updated PR:

* Adjusted naming convention & keyword-spacing
* Tweaked signatures code to work correctly in node 7.x
* Simplified `onProgress` output, unless `NODE_ENV=debug
* Introduced ignoreMask for easy file-exclusion

> The npm i on Travis is failing with Cannot find module 'array-union'. I actually saw that locally, but then I think I cleared node_modules to fix it. But clearing the cache on Travis doesn't seem to be doing the trick.

I'm struggling to reproduce that locally. I can't think of the reason other than travis cache.

dstillman commented:
OK, working after clearing the `master` cache. (Previously I only cleared the PR cache. Not sure why that didn't work.)
